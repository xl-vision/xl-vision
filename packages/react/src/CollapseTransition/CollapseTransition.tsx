import PropTypes from 'prop-types';
import React from 'react';
import { getComputedStyle, isProduction, nextFrame, onTransitionEnd } from '@xl-vision/utils';
import {
  TransitionCancelledHook,
  TransitionEndHook,
  TransitionStartHook,
  TransitionStartingHook,
  useConstantFn,
} from '@xl-vision/hooks';
import CssTransition, { CssTransitionProps } from '../CssTransition';

export interface CollapseTransitionProp
  extends Omit<CssTransitionProps, 'mountOnEnter' | 'unmountOnExit' | 'disableCss'> {
  children: React.ReactElement<React.HTMLAttributes<HTMLElement>>;
  horizontal?: boolean;
}

const CollapseTransition: React.FunctionComponent<CollapseTransitionProp> = (props) => {
  const {
    horizontal,
    onEnter,
    onEntering,
    onEntered,
    onEnterCancelled,
    onExit,
    onExiting,
    onExited,
    onExitCancelled,
    children,
    ...others
  } = props;

  const mappings = React.useMemo(() => {
    const padding1: keyof React.CSSProperties = horizontal ? 'paddingLeft' : 'paddingTop';
    const padding2: keyof React.CSSProperties = horizontal ? 'paddingRight' : 'paddingBottom';
    const size: keyof React.CSSProperties = horizontal ? 'width' : 'height';

    return {
      padding1,
      padding2,
      size,
    };
  }, [horizontal]);

  const actualSizeRef = React.useRef<string>();
  const padding1Ref = React.useRef<string>();
  const padding2Ref = React.useRef<string>();
  const sizeRef = React.useRef<string>();
  const overflowRef = React.useRef<string>();

  const handleEnter: TransitionStartHook = useConstantFn((nativeEl, transitionOnFirst) => {
    onEnter?.(nativeEl, transitionOnFirst);

    const el = nativeEl as HTMLElement;

    const { padding1, padding2, size } = mappings;

    padding1Ref.current = padding1Ref.current || el.style[padding1];
    padding2Ref.current = padding2Ref.current || el.style[padding2];
    sizeRef.current = sizeRef.current || el.style[size];
    overflowRef.current = overflowRef.current || el.style.overflow;
    actualSizeRef.current = actualSizeRef.current || getComputedStyle(nativeEl)[size];

    console.log(1, nativeEl.cloneNode(), actualSizeRef.current);

    el.style.overflow = 'hidden';
    el.style[padding1] = '0px';
    el.style[padding2] = '0px';
    el.style[size] = '0px';
    el.style.transitionDuration = '0s';

    // setTransitionStyle({
    //   overflow: 'hidden',
    //   [size]: 0,
    //   [padding1]: 0,
    //   [padding2]: 0,
    // });
    console.log(2, nativeEl.cloneNode(), actualSizeRef.current);
  });

  const handleEntering: TransitionStartingHook = useConstantFn(
    (nativeEl, done, transitionOnFirst, isCancelled) => {
      onEntering?.(nativeEl, done, transitionOnFirst, isCancelled);

      const el = nativeEl as HTMLElement;

      setTimeout(() => {
        
      })

      const { padding1, padding2, size } = mappings;
      console.log(3, nativeEl.cloneNode(), actualSizeRef.current);

      el.style[padding1] = padding1Ref.current!;
      el.style[padding2] = padding2Ref.current!;
      el.style[size] = actualSizeRef.current!;
      console.log(4, nativeEl.cloneNode(), actualSizeRef.current);

      // nextFrame(() => {
      //   if (isCancelled()) {
      //     return;
      //   }

      //   el.style[padding1] = padding1Ref.current!;
      //   el.style[padding2] = padding2Ref.current!;
      //   el.style[size] = actualSizeRef.current!;

      //   // setTransitionStyle({
      //   //   overflow: 'hidden',
      //   //   [size]: actualSizeRef.current,
      //   // });
      //   console.log(4, nativeEl.cloneNode(), actualSizeRef.current);
      //   // onTransitionEnd(el, done);
      // });
    },
  );

  const handleEntered: TransitionEndHook = useConstantFn((nativeEl, transitionOnFirst) => {
    onEntered?.(nativeEl, transitionOnFirst);
    const el = nativeEl as HTMLElement;

    // const el = nativeEl as HTMLElement;
    const { size, padding1, padding2, actualSize } = mappings;

    // el.style[size] = el.dataset[size]!;
    // el.style[padding1] = el.dataset[padding1]!;
    // el.style[padding2] = el.dataset[padding2]!;
    // el.style.overflow = el.dataset.overflow!;

    // delete el.dataset[padding1];
    // delete el.dataset[padding2];
    // delete el.dataset[size];
    // delete el.dataset.overflow;
    // delete el.dataset[actualSize];

    el.style[size] = sizeRef.current!;
    el.style.overflow = overflowRef.current!;

    actualSizeRef.current = undefined;
    padding1Ref.current = undefined;
    padding2Ref.current = undefined;
    sizeRef.current = undefined;
    overflowRef.current = undefined;
    console.log(5, nativeEl.cloneNode(), actualSizeRef.current);
  });

  const handleEnterCancelled: TransitionCancelledHook = useConstantFn(
    (nativeEl, transitionOnFirst) => {
      onEnterCancelled?.(nativeEl, transitionOnFirst);

      const el = nativeEl as HTMLElement;
      console.log('handleEnterCancelled');
    },
  );

  const handleExit: TransitionStartHook = useConstantFn((nativeEl, transitionOnFirst) => {
    onExit?.(nativeEl, transitionOnFirst);

    const { size } = mappings;

    actualSizeRef.current = actualSizeRef.current || getComputedStyle(nativeEl)[size];

    // setTransitionStyle({
    //   overflow: 'hidden',
    //   [size]: actualSizeRef.current,
    // });
  });

  const handleExiting: TransitionStartingHook = useConstantFn(
    (nativeEl, done, transitionOnFirst, isCancelled) => {
      onExiting?.(nativeEl, done, transitionOnFirst, isCancelled);

      const { padding1, padding2, size } = mappings;

      nextFrame(() => {
        if (isCancelled()) {
          return;
        }

        // setTransitionStyle({
        //   overflow: 'hidden',
        //   [padding1]: 0,
        //   [padding2]: 0,
        //   [size]: 0,
        // });
      });
    },
  );

  const handleExited: TransitionEndHook = useConstantFn((nativeEl, transitionOnFirst) => {
    onExited?.(nativeEl, transitionOnFirst);

    // setTransitionStyle({});
    actualSizeRef.current = undefined;
  });

  const handleExitCancelled: TransitionCancelledHook = useConstantFn(
    (nativeEl, transitionOnFirst) => {
      onExitCancelled?.(nativeEl, transitionOnFirst);

      const el = nativeEl as HTMLElement;
      console.log('handleExitCancelled');
    },
  );

  return (
    <CssTransition
      {...others}
      onEnter={handleEnter}
      onEntering={handleEntering}
      onEntered={handleEntered}
      onEnterCancelled={handleEnterCancelled}
      onExit={handleExit}
      onExiting={handleExiting}
      onExited={handleExited}
      onExitCancelled={handleExitCancelled}
      mountOnEnter={true}
    >
      {(show) => {
        const style = { ...(children.props as { style?: React.CSSProperties }).style };
        if (!show) {
          style.display = 'none';
        }
        return React.cloneElement(children, {
          style,
        });
      }}
    </CssTransition>
  );
};

if (!isProduction) {
  CollapseTransition.displayName = 'CollapseTransition';

  CollapseTransition.propTypes = {
    horizontal: PropTypes.bool,
  };
}

export default CollapseTransition;
