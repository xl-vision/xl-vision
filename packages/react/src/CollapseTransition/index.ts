export * from './CollapseTransition';

export { default } from './CollapseTransition';
