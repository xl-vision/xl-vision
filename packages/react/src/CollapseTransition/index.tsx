import PropTypes from 'prop-types';
import React from 'react';
import CSSTransition, { CSSTransitionClasses, TransitionElement } from '../CSSTransition';
import { forceReflow } from '../utils/transition';
import { removeClass, addClass } from '../utils/class';

export interface CollapseTransitionProp {
  children: React.ReactElement<React.HTMLAttributes<HTMLElement>>;
  unmountOnLeave?: boolean;
  in: boolean;
  transitionClasses?: CSSTransitionClasses;
  horizontal?: boolean;
  transitionOnFirst?: boolean;
}

const CollapseTransition: React.FunctionComponent<CollapseTransitionProp> = (props) => {
  const {
    children,
    transitionClasses,
    in: inProp,
    unmountOnLeave,
    horizontal,
    transitionOnFirst,
  } = props;

  const transitionEvents = React.useMemo(() => {
    const padding1 = horizontal ? 'paddingLeft' : 'paddingTop';
    const padding2 = horizontal ? 'paddingRight' : 'paddingBottom';
    const size = horizontal ? 'width' : 'height';
    const actualSize = horizontal ? 'actualWidth' : 'actualHeight';

    return {
      beforeEnter(el: TransitionElement) {
        el.dataset[padding1] = el.style[padding1];
        el.dataset[padding2] = el.style[padding2];
        el.dataset[size] = el.style[size];
        el.dataset.overflow = el.style.overflow;

        const { display } = el.dataset;

        if (display !== undefined) {
          el.style.display = display;
          el.dataset.display = undefined;
        }

        if (!el._cancelled) {
          removeClass(el, el._ctc?.enter || '');
          removeClass(el, el._ctc?.enterActive || '');
          el.dataset[actualSize] = getComputedStyle(el)[size];
        }
        el.style.overflow = 'hidden';
        el.style[size] = '0';
        el.style[padding1] = '0';
        el.style[padding2] = '0';

        if (!el._cancelled) {
          addClass(el, el._ctc?.enter || '');
          forceReflow();
          addClass(el, el._ctc?.enterActive || '');
        }
      },
      enter(el: HTMLElement) {
        el.style[size] = `${el.dataset[actualSize]!}`;
        el.style[padding1] = el.dataset[padding1]!;
        el.style[padding2] = el.dataset[padding2]!;
      },
      afterEnter(el: HTMLElement) {
        el.style[size] = el.dataset[size]!;
        el.style.overflow = el.dataset.overflow!;
      },
      enterCancelled(el: HTMLElement) {
        el.style[padding1] = el.dataset[padding1]!;
        el.style[padding2] = el.dataset[padding2]!;
        el.style[size] = el.dataset[size]!;
        el.style.overflow = el.dataset.overflow!;
      },
      beforeLeave(el: TransitionElement) {
        el.dataset[padding1] = el.style[padding1];
        el.dataset[padding2] = el.style[padding2];
        el.dataset[size] = el.style[size];
        el.dataset.overflow = el.style.overflow;

        if (!el._cancelled) {
          el.dataset[actualSize] = getComputedStyle(el)[size];
        }

        el.style[size] = `${el.dataset[actualSize]!}`;
        el.style.overflow = 'hidden';
      },
      leave(el: HTMLElement) {
        forceReflow();
        el.style[padding1] = '0';
        el.style[padding2] = '0';
        el.style[size] = '0';
      },
      afterLeave(el: HTMLElement) {
        el.style[padding1] = el.dataset[padding1]!;
        el.style[padding2] = el.dataset[padding2]!;
        el.style[size] = el.dataset[size]!;
        el.style.overflow = el.dataset.overflow!;
        el.dataset.display = el.style.display || '';
        el.style.display = 'none';
      },
      leaveCancelled(el: HTMLElement) {
        el.style[padding1] = el.dataset[padding1]!;
        el.style[padding2] = el.dataset[padding2]!;
        el.style[size] = el.dataset[size]!;
        el.style.overflow = el.dataset.overflow!;
      },
    };
  }, [horizontal]);

  return (
    <CSSTransition
      {...transitionEvents}
      transitionClasses={transitionClasses}
      transitionOnFirst={transitionOnFirst}
      in={inProp}
      mountOnEnter={true}
      unmountOnLeave={unmountOnLeave}
    >
      {children}
    </CSSTransition>
  );
};

CollapseTransition.propTypes = {
  transitionClasses: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  horizontal: PropTypes.bool,
  children: PropTypes.element.isRequired,
  in: PropTypes.bool.isRequired,
  unmountOnLeave: PropTypes.bool,
  transitionOnFirst: PropTypes.bool,
};

export default CollapseTransition;
