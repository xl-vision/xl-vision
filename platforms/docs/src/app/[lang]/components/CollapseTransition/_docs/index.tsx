'use client';

import Docs from '@xl-vision/react/CollapseTransition/__doc__/index.en-US.mdx?locale';

export default Docs;
