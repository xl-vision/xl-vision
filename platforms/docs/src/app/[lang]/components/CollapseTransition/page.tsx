import { componentRoutes } from '@docs/routes';
import createGenerateMetadata from '@docs/utils/createGenerateMetadata';
import Docs from './_docs';

export const generateMetadata = createGenerateMetadata(componentRoutes, 'CollapseTransition');

export default Docs;
