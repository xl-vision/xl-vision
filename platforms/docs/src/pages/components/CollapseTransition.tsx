import { GetStaticProps, NextPage } from 'next';
import React from 'react';
import Docs from '@xl-vision/react/CollapseTransition/__doc__/index.en-US.mdx?locale';
import { LayoutKey } from '../../layout';

const ComponentHome: NextPage = () => {
  return <Docs />;
};

export const getStaticProps: GetStaticProps<{ layout: LayoutKey }> = () => {
  return {
    props: {
      layout: 'component',
    },
  };
};

export default ComponentHome;
