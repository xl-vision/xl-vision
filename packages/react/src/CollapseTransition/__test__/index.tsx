import { mount } from 'enzyme';
import React from 'react';
import CollapseTransition from '..';

describe('CollapseTransition', () => {
  it('基本使用', () => {
    const wrapper = mount(
      <CollapseTransition in={false}>
        <div />
      </CollapseTransition>,
    );

    expect(wrapper.render()).toMatchSnapshot();

    wrapper.setProps({
      in: true,
    });

    wrapper.update();

    expect(wrapper.render()).toMatchSnapshot();
  });

  it('测试horizontal', () => {
    const wrapper = mount(
      <CollapseTransition horizontal={true} in={false}>
        <div />
      </CollapseTransition>,
    );

    expect(wrapper.render()).toMatchSnapshot();

    wrapper.setProps({
      in: true,
    });

    wrapper.update();

    expect(wrapper.render()).toMatchSnapshot();
  });
});
