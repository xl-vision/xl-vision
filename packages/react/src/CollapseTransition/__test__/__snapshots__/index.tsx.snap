// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapseTransition 基本使用 1`] = `null`;

exports[`CollapseTransition 基本使用 2`] = `
<div
  class="undefined undefined"
  data-actual-height=""
  data-height=""
  data-overflow=""
  data-padding-bottom=""
  data-padding-top=""
  style="overflow: hidden; padding-top: 0px; padding-bottom: 0px;"
/>
`;

exports[`CollapseTransition 测试horizontal 1`] = `null`;

exports[`CollapseTransition 测试horizontal 2`] = `
<div
  class="undefined undefined"
  data-actual-width=""
  data-overflow=""
  data-padding-left=""
  data-padding-right=""
  data-width=""
  style="overflow: hidden; padding-left: 0px; padding-right: 0px;"
/>
`;
